### ISSUE TITLE
OWA: Outlook.com: Unexpected end of JSON input

### ISSUE THREAD

**Issue by benbucksch** [mustang-im/mustang#215]
### Reproduction
* Set up an @outlook.com personal account
* Choose protocol OWA
* Click on Inbox
* Login

### Actual result
* Login appears to be working (not sure)
* Error: "Unexpected end of JSON input"
* Stack: 
![image](https://github.com/user-attachments/assets/023b8f55-af7f-4c8c-b94b-805885ec3ec4)

### Expected result
* No error
* Login works

### Scope
* Business accounts (neil@beonex.on...) seem to work, without the error

**Comment by jermy-c** [comment]
@benbucksch What's your `node` version? My node is `v20.16.0`. I'm wondering if it's my node version.

**Comment by benbucksch** [comment]
node 20.15.1

**Comment by jermy-c** [comment]
It might be because the body is not null but an empty string. It is normal for `JSON.parse()` to throw an error `Unexpected end of JSON input`. `JSON.parse()` doesn't throw an error for null.

**Comment by benbucksch** [comment]
I also just saw error "Unexpected token '<', "<!DOCTYPE "... is not valid JSON". Stack shows `markRead()` on an OWA account. This happened after I had the app running idle for a a day.

```stacktrace
error.ts:6 Error: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
    at message.js:153:16
    at new Promise (<anonymous>)
    at WSCall.makeCall (message.js:148:12)
    at JPCWebSocket.callRemote (protocol.js:161:31)
    at Object.fetchJSON (obj.js:198:55)
    at OWAAccount.callOWA (OWAAccount.ts:179:50)
    at OWAFolder.folderCountsChanged (OWAFolder.ts:78:37)
    at OWAFolder.listMessages (OWAFolder.ts:92:21)
    at async loadFolder (MailApp.svelte:40:28)
```

**Comment by benbucksch** [comment]
@jermy-c @NeilRashbrook Do we know *why* this error appears? I see that for both Office365 and Outlook.com accounts. I am worried that the PR #222 merely hides the fact that our push notifications do not work. We want them to work.

**Comment by NeilRashbrook** [comment]
`ev.owa2` doesn't show up on a network trace when I log in to either an Office 365 or an Outlook account, so it's quite possible that notifications don't actually work any more, although you would have to set a breakpoint in `listenForEvents` to see whether you still get any chunks (I'm not going to do that now; I was just reading through some comments to pass the time).

**Comment by NeilRashbrook** [comment]
Indeed, that endpoint now seems to be completely nonfunctional (it just returns an empty 200 OK). Unfortunately that means that the latest code now makes continual requests to the endpoint, since it doesn't actually check anything any more.

**Comment by jermy-c** [comment]
Ok, thanks for confirming that. But there must another endpoint for notifications. I've been searching for something similar in the HTTP requests but so far I've only seen this one `https://outlook.live.com/owa/0/notificationchannel/negotiate?clientProtocol=1.5&UA=0&cid=ad79233d-3af0-3eee-f53a-5e1c9d54af1a&app=Mail`. I'm still searching how it got the `cid` for that.

**Comment by NeilRashbrook** [comment]
(Also take note that this change only affects Office 365 accounts, so for example if you have an account on `exchange2019.ionos.eu` server then push notifications will still work.)

**Comment by NeilRashbrook** [comment]
On an Office 365 account I also see `https://outlook.office.com/owa/notificationchannel/negotiate` after which `/connect`, `/start`, `/abort` and `/reconnect` use the `ConnectionToken` returned by the first call, but I also can't work out where the `cid` comes from.

Interestingly, the `/connect` endpoint returns an event stream. This would normally be handled by an `EventSource` object https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events but since we're proxying requests we may have to implement this manually.

**Comment by benbucksch** [comment]
Thanks Neil and Jeremy for working on this and the insights.
@jermy-c Can you look further into this?

> the /connect endpoint returns an event stream. This would normally be handled by an EventSource object https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events but since we're proxying requests we may have to implement this manually.

We've got 3 options for the EventSource:
1. Make the EventSource call from the frontend
   * Pro: Easy
   * Con: Might not work, due to CORS. We might be lucky and CORS may not apply in this case, just like it doesn't apply to WebSockets.
   * Con: Discrepancy, between other server calls from backend and this EventSource call from frontend, might cause subtle and hard to debug bugs for some users in some setups.
2. Make the EventSource call from backend. Use EventSource implementation from node.js.
   * Con: Node.js might not offer this API
   * Con: We might need another npm module. Is it small and without any further dependencies? Is it well implemented? Is it secure?
   * Pro: Consistent with other calls
   * Pro: We don't need to implement it ourselves
   * Con: Might be difficult to get over JPC. Callbacks? Can backend call frontend?
3. Make the EventSource call from backend. Implementation it ourselves.
   * Con: We need to implement a browser API
   * Con: More code in backend
   * Con: Might be difficult to get over JPC. Callbacks? Can backend call frontend?
   * Pro: Consistent with other calls

**Comment by NeilRashbrook** [comment]
Option 1. probably won't work because we need the cookies from the partition. Option 2 doesn't seem viable since Node.js doesn't seem to provide EventSource yet and polyfills will assume a true Node environment but we need something that works with Electron. For option 3. I did find an NPM module called `eventsource-parser` which can be used to transform the output of a `TextDecoderStream` from a stream of text into a stream of events. We can then just return this stream back through JPC and the front end can use `for await` to read each event as it occurs. (We already stream text over JPC for the Exchange 2013-2019 compatible notifications).

**Comment by NeilRashbrook** [comment]
I've found the "source" of the `cid`; it's just a random string in UUID format. Here's the relevant OWA AppBoot code as pretty-printed by Firefox developer tools:
```javascript
      function r() {
        return Math.floor((1 + Math.random()) * 65536).toString(16).substring(1)
      }
      function i() {
        return r() + r() + '-' + r() + '-' + r() + '-' + r() + '-' + r() + r() + r()
      }
```

**Comment by NeilRashbrook** [comment]
The `/negotiate` endpoint works fine and gives me a connection token. Unfortunately, looking more closely at the `/connect` endpoint, it seems that it has its own authentication - the OWA front-end sends a `WWW-Authenticate: Bearer` with an access token, but I don't know where that comes from.

**Comment by jermy-c** [comment]
>  @jermy-c Can you look further into this?

Ok, will do.

**Comment by NeilRashbrook** [comment]
I found out where to get the access token, and wrote some proof of concept code in branch `neil/office-365-notifications-wip` to log the notifications to the console, but I always get HTTP status 449 insufficient client information, and the error message is unhelpful.

**Comment by jermy-c** [comment]
On a personal account, I get an HTTP 500 error.

**Comment by benbucksch** [comment]
@NeilRashbrook wrote:
> branch `neil/office-365-notifications-wip`

I merged that to master

**Comment by benbucksch** [comment]
@NeilRashbrook You wrote:
> the /connect endpoint, it seems that it has its own authentication - the OWA front-end sends a WWW-Authenticate: Bearer with an access token, but I don't know where that comes from.

Is that problem still the reason why it doesn't work, or did you overcome that problem?

**Comment by NeilRashbrook** [comment]
Ben wrote:

> did you overcome that problem?

I previously wrote:

> I found out where to get the access token

**Comment by NeilRashbrook** [comment]
I have made progress - fixing typos is essential while coding.
- `options.headers` was typo'd with a capital `H`. This didn't affect the negotation because as it turns out it didn't need any extra headers, but it did affect the connect!
- The ConnectionToken does of course need to be URI-encoded to be included in the url. (But the ConnectionId is unused.)
- There should be an `await` on the call to `callOWA` for the subscription notification request.

You weren't entirely blameless though:
- Your code to set the (unnecessary) authorisation on the negotiate call meant that authorisation wasn't being set properly on the connect options. (Not that the options would have worked anyway.)
- You were trying to use the connect options for the `start` endpoint which doesn't work. (Unfortunately I didn't notice when you made that change otherwise I would have pointed out the error at the time.)

I've pushed my work so far to branch `neil/office365-notifications-wip`.

**Comment by NeilRashbrook** [comment]
I've updated branch `neil/office365-notifications-wip` to actually process notifications, but I don't have any handling for when the stream closes yet. (Each event object generated by the backend event decoder provides an optional `retry` value, which if present should be used as a number of milliseconds delay before retrying, and an optional `id` value, which if present should be sent as the last event id header.)

**Comment by NeilRashbrook** [comment]
I still can't get `RowNotification` to work, but I have noticed that we get a `HierarchyNotification` when a message is flagged in another client, so we can at least mark the folder potentially dirty in this case.

OWA sends an `abort` when it unloads, so that's something to bear in mind for when we get around to properly supporting closing and reopening the window on the Mac.

In addition, OWA seems to send a `ping` every 5 minutes. I'm not sure what the purpose of this is for; we seem to receive notifications even without sending regular pings.

**Comment by NeilRashbrook** [comment]
The access token returned by `OWAGetAccessTokenForResource` lasts 50 minutes. As yet I can't tell whether OWA restarts when it notices that its `ping` has failed or whether the server sent events stop anyway, because for some reason they aren't showing up in Network Monitor, even though the feature is supposed to have existed since Firefox 82. OWA also tries to `abort` when the `ping` fails although that also fails of course.

**Comment by NeilRashbrook** [comment]
I left mustang running, and the server won't disconnect the stream connection for you, but it will also stop sending events after a few minutes, so it looks as if we might need to `ping` every 5 minutes (`TransportConnectTimeout`?) to keep the notifications alive, but also every 40 minutes (`KeepAliveTimeout`?) we need to abort the connection and start a new one.

**Comment by benbucksch** [comment]
Take a look at class `logic/Auth/OAuth2`, how I did the refresh timer there.

It makes sense to have a separate class `OWANotification` for the notifications, which then calls OWAAccount. The timer can then be within that class.

``OWANotification.onHierarchyNotification()` (rename handleFoo to onFoo, please) could be in that new class. This way, the reading of the result JSON stays in the same place. If you find that you need protected functions or properties of `OWAAccount`, you can make higher level functions in `OWAAccount` as appropriate.

**Comment by NeilRashbrook** [comment]
I can confirm that sending regular pings is needed to receive events after the first five minutes and also that sending an abort will end the connection on the server allowing us to restart the connection.

I did my best to address your review comments in branch `neil/owa-notifications-class` as a further commit after the changes in `neil/office365-notifications-wip`.

**Comment by benbucksch** [comment]
Thank you! Glad that it starts working.

I think you forgot to push the branch.

So, three main problem why didn't work was simply typos etc. while coding?
What is working now, and what isn't?

**Comment by NeilRashbrook** [comment]
According to my `git` command line, both branches are up-to-date on the server. On the other hand, I do see a branch `neil/office-365-notifications-wip`; I don't remember what happened there, but that's the wrong branch, and has already been merged to master.

The functionality on `neil/office365-notifications-wip` and `neil/owa-notifications-class` is the same as provided on previous versions of Exchange in that new mail is automatically fetched and if other folders are changed on the server while mustang is open then when you open that folder in mustang it will always update it, although I think I may have implemented that incorrectly and it will always update any folder the first time you open it in mustang whereas before it would always check the message counts first.

**Comment by benbucksch** [comment]
I've created pull request #330 based on branch `neil/owa-notifications-class`, created files for each class, tweaked the caller a little bit, and merged it to master.

**Comment by benbucksch** [comment]
@NeilRashbrook
On startup, with neil@be (Office365) as OWA account, I am getting the error:
```
startup.ts:80 OWAError: HTTP 500 
    at OWAAccount.callOWA (OWAAccount.ts:208:13)
    at async OWAOffice365Notifications.start (OWAOffice365Notifications.ts:16:24)
```

**Comment by benbucksch** [comment]
> I may have implemented that incorrectly and it will always update any folder the first time you open it in mustang, whereas before it would always check the message counts first.

@NeilRashbrook Yes, indeed, your change also affected all other cases that call `folder.fromJSON()`. I've reverted this change in git commit a0259186 , and added this `folder.dirty = true` in `handleHierarchyNotification()`. Can you please check whether this commit is correct?

**Comment by NeilRashbrook** [comment]
`git diff a025918~2 a025918 **/OWAFolder.ts` produces no output which seems good to me.

**Comment by benbucksch** [comment]
@NeilRashbrook I meant whether my change is _logically_ correct. I did change something. See my last message describing it.

**Comment by NeilRashbrook** [comment]
You correctly reverted my change that made the folder dirty in too many cases and limited it to just hierarchy notifications.

**Comment by benbucksch** [comment]
Super, thanks!
